File: app/src/main/java/com/marverenic/music/utils/BugsnagTree.java
package com.marverenic.music.utils;

import android.app.ActivityManager;
import android.content.Context;

import com.bugsnag.android.Bugsnag;
import com.bugsnag.android.Error;
import com.bugsnag.android.Severity;

import org.jetbrains.annotations.Nullable;

import java.util.ArrayDeque;

import timber.log.Timber;

import static android.util.Log.DEBUG;
import static android.util.Log.ERROR;
import static android.util.Log.INFO;
import static android.util.Log.VERBOSE;
import static android.util.Log.WARN;
import static java.util.Locale.US;

public class BugsnagTree extends Timber.Tree {

    private static final int BUFFER_SIZE = 200;

    private final String processName;
    private final ArrayDeque<String> logs = new ArrayDeque<>(BUFFER_SIZE + 1);

    public BugsnagTree(Context context) {
        processName = currentProcessName(context);
    }

    @Override
    protected void log(int priority, @Nullable String tag, @Nullable String message,
                       @Nullable Throwable throwable) {
        if (priority == VERBOSE || priority == DEBUG) {
            return;
        }

        if (message != null) {
            String log = System.currentTimeMillis() + " "
                    + getPrioritySymbol(priority) + ": " + message;

            synchronized (logs) {
                logs.add(log);
                if (logs.size() > BUFFER_SIZE) {
                    logs.removeFirst();
                }
            }
        }

        if (throwable != null) {
            Severity severity;
            if (priority == ERROR) {
                severity = Severity.ERROR;
            } else if (priority == WARN) {
                severity = Severity.WARNING;
            } else if (priority == INFO) {
                severity = Severity.INFO;
            } else {
                return;
            }

            Bugsnag.notify(throwable, report -> {
                Error error = report.getError();
                error.setSeverity(severity);
                error.addToTab("App", "processName", processName);
                synchronized (logs) {
                    int i = 0;
                    for (String log : logs) {
                        error.addToTab("Log", String.format(US, "%03d", i++), log);
                    }
                }
            });
        }
    }

    private String getPrioritySymbol(int priority) {
        switch (priority) {
            case ERROR: return "E";
            case WARN: return "W";
            case INFO: return "I";
            case DEBUG: return "D";
            case VERBOSE: return "V";
            default: return Integer.toString(priority);
        }
    }

    private static String currentProcessName(Context context) {
        int pid = android.os.Process.myPid();
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningAppProcessInfo processInfo : manager.getRunningAppProcesses()) {
            if (processInfo.pid == pid) {
                return processInfo.processName;
            }
        }
        return null;
    }
}

File: app/src/main/java/com/marverenic/music/utils/BugsnagTree.kt
package com.marverenic.music.utils

import android.app.ActivityManager
import android.content.Context
import android.util.Log.*
import androidx.core.content.getSystemService
import com.bugsnag.android.Bugsnag
import com.bugsnag.android.Severity
import timber.log.Timber
import java.util.*
import java.util.Locale.US

class BugsnagTree(
    context: Context,
    private val bufferSize: Int = 200
) : Timber.Tree() {

    private val processName = context.currentProcessName()
    private val logs = ArrayDeque<String>(bufferSize + 1)

    override fun log(
        priority: Int,
        tag: String?,
        message: String,
        throwable: Throwable?
    ) {
        if (priority != VERBOSE && priority != DEBUG) {
            addLog(message, priority)
            throwable?.let { logThrowable(it, priority) }
        }
    }

    private fun addLog(message: String, priority: Int) {
        val log = "${System.currentTimeMillis()} ${priority.toSeverityString()}: $message"

        synchronized(logs) {
            logs.add(log)
            if (logs.size > bufferSize) {
                logs.removeFirst()
            }
        }
    }

    private fun logThrowable(throwable: Throwable, priority: Int) {
        val severity = when (priority) {
            ERROR -> Severity.ERROR
            WARN -> Severity.WARNING
            INFO -> Severity.INFO
            else -> return
        }

        Bugsnag.notify(throwable) { report ->
            val error = report.error ?: return@notify

            error.severity = severity
            error.addToTab("App", "processName", processName)
            synchronized(logs) {
                logs.forEachIndexed { index, log ->
                    error.addToTab("Log", String.format(US, "%03d", index), log)
                }
            }
        }
    }

    private fun Int.toSeverityString(): String {
        return when (this) {
            ERROR -> "E"
            WARN -> "W"
            INFO -> "I"
            DEBUG -> "D"
            VERBOSE -> "V"
            else -> toString()
        }
    }

    private fun Context.currentProcessName(): String? {
        val pid = android.os.Process.myPid()
        return getSystemService<ActivityManager>()
            ?.runningAppProcesses
            ?.firstOrNull { it.pid == pid }
            ?.processName
    }
}
